File: typescript/packages/core/src/policy/integrity-tracker.ts
/**
 * Tool Integrity Tracker.
 *
 * Tracks SHA-256 hashes of tool YAML content to detect modifications.
 * Used during hot-reload to decide whether re-validation is needed.
 */

import { createHash } from 'crypto';

export interface IntegrityRecord {
  hash: string;
  source: 'trusted' | 'untrusted';
  validatedAt: Date;
}

export type IntegrityAction =
  | { action: 'keep'; reason: 'unchanged' }
  | { action: 'revalidate'; reason: 'content-modified' }
  | { action: 'revalidate'; reason: 'source-changed' }
  | { action: 'validate'; reason: 'new-tool' };

export class ToolIntegrityTracker {
  private readonly records: Map<string, IntegrityRecord> = new Map();

  /**
   * Compute SHA-256 hash of content.
   */
  computeHash(content: string): string {
    return createHash('sha256').update(content, 'utf8').digest('hex');
  }

  /**
   * Called when a tool is loaded. Compares the content hash and source with the stored
   * record and returns what action the caller should take.
   * If the tool moved between trusted/untrusted paths, revalidate is required even if content unchanged.
   */
  onToolLoaded(
    toolName: string,
    yamlContent: string,
    source: 'trusted' | 'untrusted'
  ): IntegrityAction {
    const hash = this.computeHash(yamlContent);
    const existing = this.records.get(toolName);

    if (!existing) {
      return { action: 'validate', reason: 'new-tool' };
    }

    // Check if source changed (trusted ↔ untrusted): must revalidate for policy enforcement
    if (existing.source !== source) {
      return { action: 'revalidate', reason: 'source-changed' };
    }

    if (existing.hash === hash) {
      return { action: 'keep', reason: 'unchanged' };
    }

    return { action: 'revalidate', reason: 'content-modified' };
  }

  /**
   * Record a tool's hash after successful validation/loading.
   */
  record(toolName: string, yamlContent: string, source: 'trusted' | 'untrusted'): void {
    this.records.set(toolName, {
      hash: this.computeHash(yamlContent),
      source,
      validatedAt: new Date(),
    });
  }

  /**
   * Get the stored record for a tool.
   */
  getRecord(toolName: string): IntegrityRecord | undefined {
    return this.records.get(toolName);
  }

  /**
   * Get the stored hash for a tool.
   */
  getHash(toolName: string): string | undefined {
    return this.records.get(toolName)?.hash;
  }

  /**
   * Remove a tool entry (e.g. when it's been removed from disk).
   */
  removeEntry(toolName: string): boolean {
    return this.records.delete(toolName);
  }

  /**
   * Clear all records. Used when doing a full reset.
   */
  clear(): void {
    this.records.clear();
  }

  /**
   * Get the number of tracked tools.
   */
  get size(): number {
    return this.records.size;
  }
}

File: python/packages/core/src/matimo/policy/integrity_tracker.py
"""
Integrity tracker — SHA-256 fingerprinting for tool hot-reload.
Mirrors: packages/core/src/policy/integrity-tracker.ts
"""
from __future__ import annotations

import hashlib
import logging
from enum import Enum
from pathlib import Path

logger = logging.getLogger("matimo")


class IntegrityAction(str, Enum):
    KEEP = "keep"           # content unchanged — skip re-validation
    REVALIDATE = "revalidate"  # content or source changed — re-validate
    VALIDATE = "validate"   # new tool — run full validation


class ToolIntegrityTracker:
    """
    Tracks SHA-256 hashes of loaded tool YAML files to determine whether
    re-validation is needed on hot-reload.
    Mirrors: ToolIntegrityTracker in integrity-tracker.ts
    """

    def __init__(self) -> None:
        # tool_name → (hash, source_path)
        self._hashes: dict[str, tuple[str, str]] = {}

    def get_action(self, tool_name: str, file_path: str) -> IntegrityAction:
        """
        Determine the action for a tool during reload.

        Returns:
            KEEP       — file unchanged since last load
            REVALIDATE — file content or path changed
            VALIDATE   — tool not previously seen
        """
        current_hash = self._hash_file(file_path)
        if current_hash is None:
            return IntegrityAction.VALIDATE

        if tool_name not in self._hashes:
            return IntegrityAction.VALIDATE

        stored_hash, stored_path = self._hashes[tool_name]

        if stored_path != file_path:
            logger.debug(
                "Tool '%s' source path changed: %s → %s", tool_name, stored_path, file_path
            )
            return IntegrityAction.REVALIDATE

        if stored_hash != current_hash:
            logger.debug("Tool '%s' content changed (hash mismatch)", tool_name)
            return IntegrityAction.REVALIDATE

        return IntegrityAction.KEEP

    def record(self, tool_name: str, file_path: str) -> str | None:
        """
        Record the current hash for a tool definition file.
        Returns the hash string, or None if the file cannot be read.
        """
        h = self._hash_file(file_path)
        if h is not None:
            self._hashes[tool_name] = (h, file_path)
        return h

    def remove(self, tool_name: str) -> None:
        self._hashes.pop(tool_name, None)

    def hash_for(self, tool_name: str) -> str | None:
        entry = self._hashes.get(tool_name)
        return entry[0] if entry else None

    def clear(self) -> None:
        self._hashes.clear()

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    @staticmethod
    def _hash_file(file_path: str) -> str | None:
        try:
            content = Path(file_path).read_bytes()
            return hashlib.sha256(content).hexdigest()
        except OSError:
            return None
